aquadoggo/src/domain.rs

```diff
@@ -0,0 +1,1060 @@
+
+            // Get the document_id for the document_view_id contained in previous operations.
+            // This performs several validation steps (check method doc string).
+            let document_id =
+                get_validate_document_id_for_view_id(store, &previous_operations).await?;
+
+            // Ensure the document isn't deleted.
+            ensure_document_not_deleted(store, &document_id)
+                .await
+                .map_err(|_| {
+                    "You are trying to update or delete a document which has been deleted"
+                })?;
+
+            document_id
+        }
+    };
+
+    // Verify the claimed log id against the expected one for this document id and author.
+    verify_log_id(store, &author, log_id, &document_id).await?;
+
+    ///////////////
+    // STORE LOG //
+    ///////////////
+
+    // If this is a CREATE operation it goes into a new log which we insert here.
+    if operation.is_create() {
+        let log = S::StorageLog::new(&author, &operation.schema(), &document_id, log_id);
+
+        store.insert_log(log).await?;
+    }
+
+    ///////////////////////////////
+    // STORE ENTRY AND OPERATION //
+    ///////////////////////////////
+
+    // Insert the entry into the store.
+    store
+        .insert_entry(S::StorageEntry::new(entry_encoded, operation_encoded)?)
+        .await?;
+    // Insert the operation into the store.
+    store
+        .insert_operation(
+            &S::StorageOperation::new(&author, &entry_encoded.hash().into(), &operation).unwrap(),
+            &document_id,
+        )
+        .await?;
+
+    /////////////////////////////////////
+    // DETERMINE NEXT ENTRY ARG VALUES //
+    /////////////////////////////////////
+
+    let next_seq_num = increment_seq_num(&mut seq_num.clone())?;
+    let backlink = Some(entry_encoded.hash());
+
+    // Check if skiplink is required and return hash if so
+    let skiplink = if is_lipmaa_required(next_seq_num.as_u64()) {
+        let skiplink_seq_num = next_seq_num.skiplink_seq_num().unwrap();
+        match store
+            .get_entry_at_seq_num(&author, log_id, &skiplink_seq_num)
+            .await?
+        {
+            Some(entry) => Ok(Some(entry)),
+            None => Err(anyhow!("Expected next skiplink missing")),
+        }
+    } else {
+        Ok(None)
+    }?
+    .map(|entry| entry.hash());
+
+    Ok(NextEntryArguments {
+        log_id: (*log_id).into(),
+        seq_num: next_seq_num.into(),
+        backlink: backlink.map(|hash| hash.into()),
+        skiplink: skiplink.map(|hash| hash.into()),
+    })
+}
+
+/// Attempt to identify the document id for view id contained in a `next_args` request. This will fail if:
+/// - any of the operations contained in the view id _don't_ exist in the store
+/// - any of the operations contained in the view id return a different document id than any of the others
+pub async fn get_validate_document_id_for_view_id<S: StorageProvider>(
```

Comment:
This method name "get_validate_document_id_for_view_id" is slightly confusing, maybe add a "and" after "get"? Or remove "validate"? So it would read: "get_and_validate_document_id_for_view_id" or "get_document_id_for_view_id".

Reply:
`get_checked_...` ?

Reply:
(I just learned about the use of "checked" [in the std lib](https://doc.rust-lang.org/std/?search=checked) from [this interesting post](https://blog.yoshuawuyts.com/unsafe-syntax/))

Reply:
I like it!